python/pyarrow/table.pxi

```diff
@@ -2387,3 +2429,60 @@ def _from_pydict(cls, mapping, schema, metadata):
         return cls.from_arrays(arrays, schema=schema, metadata=metadata)
     else:
         raise TypeError('Schema must be an instance of pyarrow.Schema')
+
+
+class TableGroupBy:
+    """
+    A grouping of columns in a table on which to perform aggregations.
+    """
+
+    def __init__(self, table, keys):
+        if isinstance(keys, str):
+            keys = [keys]
+
+        self._table = table
+        self.keys = keys
+
+    def aggregate(self, aggregations):
+        """
+        Perform an aggregation over the grouped columns of the table.
+
+        Parameters
+        ----------
+        aggregations : list[tuple(str, str)] or\
+                       list[tuple(str, str, FunctionOptions)]
+            List of tuples made of aggregation functions names followed
+            by column names and optionally aggregation function options.
```

Comment:
Would it make more sense to do this the other way around? 
Now you have `("func", "col", options)`, but so `("col", "func", options)`.

This way the function name and options are closer together (if the options are specified). That order also matches the resulting name (which is "col_func"). And it also matches the order in pandas I think.

Reply:
Also, I think it would be good to give a small example here (just to see it visually, something like ```for example ``[("col_A", "sum"), ("col_B", "min", options)]`` ```

Reply:
+1 on changing the order. It's `('col', 'func')` in pandas.

Reply:
Well, in pandas it's reversed because the aggregations are keyed by the column `"col": ["sum", "count"]`, here the aggregations are just listed so there is no benefit in having the column first and I felt it was more "natural" if you read "sum X" our "count X" than "X count" etc...

Anyway I don't have a strong opinion so if you feel it's better to reverse them I'm ok with that

Reply:
reversed order of arguments and added example 👍

Reply:
> Well, in pandas it's reversed because the aggregations are keyed by the column

That's for the interface when specifying a dictionary. But in pandas you can also do something like `.agg(b_sum=('b', 'sum'))`, in which case it is a bit more similar as here.